src/creatures/players/player.cpp

```diff
@@ -814,14 +823,23 @@ bool Player::canWalkthroughEx(const Creature* creature) const
 	}
 
 	const Player* player = creature->getPlayer();
+	// Lua npcs
 	const Npc* npc = creature->getNpc();
+	// XML npcs
+	const NpcOld* npcOld = creature->getNpcOld();
 	if (player) {
 		const Tile* playerTile = player->getTile();
 		return playerTile && (playerTile->hasFlag(TILESTATE_NOPVPZONE) || playerTile->hasFlag(TILESTATE_PROTECTIONZONE) || player->getLevel() <= static_cast<uint32_t>(g_config.getNumber(PROTECTION_LEVEL)) || g_game.getWorldType() == WORLD_TYPE_NO_PVP);
+	// Lua npcs
 	} else if (npc) {
-		const Tile* tile = npc->getTile();
-		const HouseTile* houseTile = dynamic_cast<const HouseTile*>(tile);
-		return (houseTile != nullptr);
+		const Tile* npcTile = npc->getTile();
+		const HouseTile* npcHouseTile = dynamic_cast<const HouseTile*>(npcTile);
+		return (npcHouseTile != nullptr);
+	// XML npcs
+	} else if (npcOld) {
+		const Tile* npcOldTile = npcOld->getTile();
+		const HouseTile* npcOldHouseTile = dynamic_cast<const HouseTile*>(npcOldTile);
+		return (npcOldHouseTile != nullptr);
```

Comment:
same as above, you should have a single getNpc that returns a NPC interface and the creature method itself understand if it should return the lua or the xml npc.